US-12: Indoor shortest travel route

bit172 opened the pull request:
This PR implements #43.

It shows the indoor route and simple directions for any POI or building that has a floor plan.
There is an accessibility option. To test the accessibility option, a route from the Security POI to the Police St. Exit POI on the MB building can be used.

osahtout committed:
Add travelRoute file #43

osahtout committed:
Path shown based on chosen floor #43

osahtout committed:
Began travel steps and handled marker location to POI #43

osahtout committed:
supercheck #43

osahtout committed:
Added travel steps text and refactored travel route logic #43

osahtout committed:
Added path planning service and reorg files #43

bit172 committed:
Refactor TravelRoute #43
by moving the animations logic to the MapScreen component

bit172 committed:
Improve Directions Text, add path-planning service  #43
fix small bug in path finding service

bit172 committed:
Handle current location as start #43
Since we currently do not have outdoor directions, the start of the user
location is the entrance of the destination building

osahtout committed:
Add docs and check building without floor plans #43

bit172 committed:
Fix check for building plans #43

osahtout committed:
Added tests and renamed files #43

osahtout committed:
Rebased after 44 merge and fix tests #43

bit172 committed:
Add accesibility option #43

bit172 committed:
Fix styling for omnibox travel mode buttons #43
Centered the colored background of icons

RGPosadas commented on __tests__/__testData__/indoor-travel-route.ts:
```diff
@@ -0,0 +1,279 @@
+import { POICategory } from "../../src/types/main";
+
+export const mockPOIs = [
```
Missing type declarations on the exported constants

r-saba replied:
Possible to add types to this entire file?

RGPosadas commented on __tests__/indoor-travel-route.test.js:
```diff
@@ -0,0 +1,36 @@
+    expect(tree).toMatchSnapshot();
+  });
+
+  it("it should alert the user if theirs no indoor floor plans", () => {
```
Typo: 
`it("it should alert the user if theirs no indoor floor plans"`
to
`it("should alert the user if there is no indoor floor plans"`

RGPosadas commented on src/components/travel-route/indoor-travel-route.component.tsx:
```diff
@@ -0,0 +1,75 @@
+ * @param end
+ * @param chosenFloorLevel
+ * @param animateToStartLocation
+ */
```
Your param descriptions need to be updated

RGPosadas commented on src/services/indoor-path-planning.service.ts:
```diff
@@ -0,0 +1,184 @@
+    endLocation: MarkerLocation,
+    accessible: boolean,
+    chosenFloorLevel: number
+  ): Line[][] => {
```
Param docs need to be updated for this whole file

RGPosadas reviewed (commented):
#### Main Concern
1. Taking a route from H9 to MB, it tells you to take the elavator with kind of confusing directions (do I go up or down? do I exit the building?)
![image](https://user-images.githubusercontent.com/33991121/79171218-7c8c0c80-7dbf-11ea-89e7-802c7554e7b5.png)

2. I am already in the building, however I am told to enter the building again.
![image](https://user-images.githubusercontent.com/33991121/79171309-ac3b1480-7dbf-11ea-8be5-a9767d57973e.png)

3. Wouldn't taking EscalatorUp on the 9th floor lead me to the 10th floor?
![image](https://user-images.githubusercontent.com/33991121/79171423-fcb27200-7dbf-11ea-928f-9f992c8e63a4.png)


#### Individual Comments
Left some individual, non-blocking comments

r-saba commented on __tests__/indoor-directions-steps.test.js:
```diff
@@ -0,0 +1,50 @@
+);
+
+describe("Indoor Directions Text component", () => {
+  it("should match indoor travel directions steps snapshot", () => {
```
Did you mean Indoor Directions `Steps` component?

r-saba commented on __tests__/indoor-travel-route.test.js:
```diff
@@ -0,0 +1,36 @@
+import IndoorTravelRoute from "../src/components/travel-route/indoor-travel-route.component";
+import renderer from "react-test-renderer";
+import { shallow, mount } from "enzyme";
+import { Alert } from "react-native";
```
unused import

r-saba commented on src/components/search/start-travel.component.tsx:
```diff
@@ -12,19 +12,21 @@ import { TravelState } from "../../types/main";
  * @param onFloorPickerButtonPress Function that handles floorpicker press event
  */
-const StartTravel = ({ setTravelState }: IProps) => {
+const StartTravel = ({ setTravelState, onStartTravelPlan }: IProps) => {
   return (
```
make sure to update the params for this function.

r-saba commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -85,6 +91,10 @@ const OmniboxDirections = ({
   updateSearchResults,
   startLocationDisplay,
   setStartLocationDisplay,
+  travelState,
+  travelMode,
+  setTravelMode,
+  onStartTravelPlan,
```
forgot to add some of these to the params in docs of the function

r-saba commented on src/components/travel-route/indoor-directions-steps.component.tsx:
```diff
@@ -0,0 +1,154 @@
+/**
+ * A sliding panel that shows the steps to the destination
+ */
+class IndoorDirectionSteps extends React.Component<IProps, IState> {
+  _panel: SlidingUpPanel;
+
+  _draggedValue: Animated.Value;
+
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      allowDragging: true,
+    };
+  }
```
Oh man been a while since I've seen a react class - Not a blocker or change request, just a comment :D

r-saba commented on src/components/travel-route/indoor-directions-steps.component.tsx:
```diff
@@ -0,0 +1,154 @@
+          animatedValue={this._draggedValue}
+          showBackdrop={false}
+          ref={(c) => (this._panel = c)}
+          // Speed of the panel
```
Good and helpful comment - Not a blocker or a change request

r-saba commented on src/components/travel-route/indoor-directions-steps.component.tsx:
```diff
@@ -0,0 +1,154 @@
+                onTouchStart={() => this.setState({ allowDragging: false })}
+                onTouchEnd={() => this.setState({ allowDragging: true })}
+                onTouchCancel={() => this.setState({ allowDragging: true })}
+                style={{ width: "100%" }}
```
inline style should be moved to the stylesheet below

r-saba commented on src/components/travel-route/indoor-travel-route.component.tsx:
```diff
@@ -0,0 +1,75 @@
+ * @param chosenFloorLevel
+ * @param animateToStartLocation
+ */
+const IndoorTravelRoute = ({
+  travelMode,
+  start,
+  end,
+  chosenFloorLevel,
+  setTravelState,
```
This component is so easy to read, great! - Not a blocker or a change request

ys-lin commented on __tests__/indoor-directions-steps.test.js:
```diff
@@ -0,0 +1,50 @@
+const mockFloorPath = PathFindingService.getInstance().findPathBetweenPOIs(
+  testData.mockPOIs[0],
+  testData.mockPOIs[1]
+);
```
findPathBetweenPOIs takes accessibilityMode: boolean, does that default to false?

r-saba commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -246,19 +275,39 @@ const MapScreen = () => {
+    );
+  };
+
+  const isTravelling = (): boolean => {
+    return travelState === TravelState.TRAVELLING;
+  };
+  const isPlanning = (): boolean => {
+    return travelState === TravelState.PLANNING;
+  };
+
+  const isTravelStateNone = (): boolean => {
+    return travelState === TravelState.NONE;
+  };
```
Thanks for doing this but now the burden is on you to ensure it's used everywhere :(. Please find and replace everywhere that stil uses the old method to check the travelstate.

osahtout replied:
Because we have those comparisons all over the map-screen, I wanted to centralize them and have one method so it can be clear to others. It's not meant to be exported or passed to other component since it will break test.

r-saba replied:
Disagree, we shouldn't be checking it one way in one component and another way in another. It's also only used in one other component. I don't think this is a big enough reason to block though

JimLau49 commented on src/screens/map-screen/map-screen.component.tsx:
```diff
@@ -150,6 +178,7 @@ const MapScreen = () => {
    * @param index Index of the active floor
    */
   const onFloorPickerButtonPress = (index: number) => {
+    setFloorLevel(9 - index);
```
9 could be moved to constants

JimLau49 commented on src/components/search/start-travel.component.tsx:
```diff
@@ -12,19 +12,21 @@ import { TravelState } from "../../types/main";
 
 /**
  * Component for controlling indoor floors when in the indoor view.
  * @param onFloorPickerButtonPress Function that handles floorpicker press event
```
I don't see onFloorPickerButtonPress being used anywhere in this function

r-saba reviewed (changes requested):
Very cool PR and I see initiative was taken to make the complex files easier to read so thank you and it helped! Left minor comments.

JimLau49 commented on __tests__/start-travel.test.js:
```diff
@@ -12,12 +12,17 @@ describe("StartTravel component", () => {
 
   it("should call provided function on press", () => {
```
Could be more specific about what the function is and what it does

JimLau49 commented on src/components/search/omnibox-directions.component.tsx:
```diff
@@ -247,14 +257,32 @@ const OmniboxDirections = ({
               <FontAwesome
                 name="wheelchair"
-                size={24}
+                size={26}
```
Could be moved to a constant

bit172 replied:
I will not move this since this is not part of my ticket. I only changed this size to make the wheelchair Icon look bigger.

bit172 committed:
Fix directions text #43

JimLau49 reviewed (commented):
Commented some small changes, if time permits i'd like to see addressed.
Otherwise I will approve when I see the other changes addressed 

### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [ ] All acceptance tests pass (N/A since I cannot run e2e tests)
- [x] If ticket is a user story, there must exist an automated acceptance test for it

### Acceptance Criteria
- [x] Route shown is fastest and shortest
- [x] I can see the route on my map
- [x] Have indoor to indoor routing
- [x] Must follow [DoD](https://github.com/RGPosadas/WayFinder/wiki/Definition-of-Done)
- [x] Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=392%3A2)

### USER FLOW ACCEPTANCE TEST
- [x] 1. Once starting location and destination room are chosen, a route is given to the user
- [x] 2. Route must be shortest and fastest path
- [x] 3. Route is clear on the map
- [x] 4. Navigation instructions are simple (such as: go up to 9th floor). No left/right turns, etc.

ys-lin committed:
address comments in the PR #43

- fix failed test and add types to testdata
- fix typo and remove unused imports

bit172 committed:
Update snapshot and directions text #43
Directions on the same floor will only give one step

r-saba reviewed (dismissed):
Approved considering the requested code changes are minor and don't affect the US/behaviour.

### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [ ] All acceptance tests pass (N/A since I cannot run e2e tests)
- [x] If ticket is a user story, there must exist an automated acceptance test for it

### Acceptance Criteria
- [x] Route shown is fastest and shortest
- [x] I can see the route on my map
- [x] Have indoor to indoor routing
- [x] Must follow [DoD](https://github.com/RGPosadas/WayFinder/wiki/Definition-of-Done)
- [x] Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=392%3A2)

### USER FLOW ACCEPTANCE TEST
- [x] 1. Once starting location and destination room are chosen, a route is given to the user
- [x] 2. Route must be shortest and fastest path
- [x] 3. Route is clear on the map
- [x] 4. Navigation instructions are simple (such as: go up to 9th floor). No left/right turns, etc.

bit172 committed:
address pr comments and change requests #43

ritchellesap committed:
add US-12 and US-13 acceptance tests #43

Also modified testIDs, updated failing e2e tests with the new implementation, and updated snapshots

RGPosadas reviewed (approved):
### DoD
- [x] Documentation present for every component and function
- [x] Unit and snapshot tests are implemented for every component and function
- [x] All unit tests pass
- [x] Implementation matches the corresponding mockup from Figma
- [x] All acceptance tests pass
- [x] If ticket is a user story, there must exist an automated acceptance test for it

### Acceptance Criteria
- [x] Route shown is fastest and shortest
- [x] I can see the route on my map
- [x] Have indoor to indoor routing
- [x] Must match [Mockup](https://www.figma.com/file/1XeyBVJ2XyC6AHDHYwxvdG/WayFinder-SOEN390?node-id=392%3A2)

### USER FLOW ACCEPTANCE TEST
- [x] 1. Once starting location and destination room are chosen, a route is given to the user
- [x] 2. Route must be shortest and fastest path
- [x] 3. Route is clear on the map
- [x] 4. Navigation instructions are simple (such as: go up to 9th floor). No left/right turns, etc.

### Notes
Good job everyone! Amazing implementation, thank you for your hard work 😄
